### PR TITLE
Ratings with game modes

### PR DESCRIPTION
**Pull request by 48klocs**
Along with the sandbox reset for the Go Fast update, the folks behind the DTR API plumbed in another feature into the D2 side of things that I've been wanting - game modes.

The values for the modes are based off of DestinyActivityModeType (see https://bungie-net.github.io/multi/schema_Destiny-HistoricalStats-Definitions-DestinyActivityModeType.html#schema_Destiny-HistoricalStats-Definitions-DestinyActivityModeType for the values). For our purposes, we're starting with PvE, PvP, raid (can only call out the one) and To9.

There's a new option in the settings menu to select it (it defaults to not specified, which pulls all the reviews). Reviews now mention which game mode they were for.

ratings-with-game-modes

I also added an option to the header to toggle between the modes, because going to the settings menu every time is a bit too much work.

game-mode-quick-switch

### CODE REVIEW

**Comment by delphiactual** [comment]
|Mode | DestinyActivityModeDefinition.name |
|--|--|
| PVE | 1164760493 |
| Crucible | 1164760504 |
| Raid | 2043403989 |
| Trials | 1370326378 |

Then you only need the `Not Applicable` mode key to be translated or alternatively you could use Vendors.All

**Comment by duckmanBR** [comment]
And you could even display the icon for each mode on the reviews when "not specified" is selected.

**Comment by kyleshay** [comment]
Yeeeeah!! This is awesome @48klocs ^_^ 

Few things...

- [ ] Could we get all of the text reviews (or is that how it currently works?) and have a filter to show/hide them on the move box? That way you could click on an item, see that it sucks for trials... but realize from someones review it's great in PvE? 

- [ ] I know it would require an API side change (could be done as another PR), but I wonder if it would help us if we could all of the ratings for an item for an item hash... so that we could easily toggle between modes w/o the addl request. The main benefit is that we could wire that up to the search filter, and then a user could say something like `pvprating>4.5`, regardless of what mode they had selected.
```
[{ 
  hash: 12345678,
  ratings: {
    all: 3.1,
    trials: 1.0, 
    pvp: 2.4, 
    pve: 4.9, 
    raid: 4.0
  }
}]
```

**Comment by kyleshay** [comment]
If we do keep that 👍 dropdown, we may also want to consider add a header to that popup that says something like "Item rating displayed/shown on item icon", so that it gives some more context about what it's for?

**Comment by 48klocs** [comment]
@delphiactual thanks for the hash IDs - got them tied in. The unspecified game mode in the manifest was too wordy, so we'll go with translating the two (not specified + unknown mode in case someone stuffs some really bogus data in DTR).

@duckmanBR I didn't try icons, because my gut feeling was that they would be more noise than signal on the reviews page. You can feel free to fool around with it when this lands to see.

@kyleshay currently, every different mode (including the "all modes" (unspecified mode) mode) is a separate fetch / ratings call. We could do more work to cache these separately on our side, but the API is generally pretty performant, so I didn't see much value in keying the cache up at this point.

Having a `pvprating` search option would be pretty sweet, but it would also take some fairly complex doing, since it goes beyond looking at data on items in our stores and on to realizing that we're missing data and loading it and not mutating the items we have loaded (think: the case where you've selected PvE ratings to be shown and then search for `pvprating`).

File: src/app/item-review/item-review.component.js [hunk 1]
```diff
@@ -2,8 +2,9 @@ import _ from 'underscore';
 import { settings } from '../settings/settings';
 import template from './item-review.html';
 import './item-review.scss';
+import { getDefinitions } from '../destiny2/d2-definitions.service';
```
[the remark this answers is not in the document]
Comment (48klocs) [reply src/app/item-review/item-review.component.js]: Is Hound getting tripped up because this is a TS file being referenced from a JS file?

File: src/app/destiny2/d2-definitions.service.ts [hunk 2]
```diff
@@ -127,6 +132,9 @@ function getDefinitionsUncached(): IPromise<D2ManifestDefinitions> {
         defs[tableShort] = D2ManifestService.getAllRecords(db, table);
       });
 
+      const eagerTablesLoaded = 'eagerTablesLoaded';
+      defs[eagerTablesLoaded] = true;
```
Comment (bhollis) [inline src/app/destiny2/d2-definitions.service.ts]: What's this for, and why can't it just be `defs.eagerTablesLoaded = true`?

File: src/app/destinyTrackerApi/d2-bulkFetcher.ts [hunk 3]
```diff
@@ -20,16 +20,16 @@ class D2BulkFetcher {
     this._reviewDataCache = reviewDataCache;
   }
 
-  _getBulkWeaponDataEndpointPost(itemList: DtrItem[], platformSelection: number) {
+  _getBulkWeaponDataEndpointPost(itemList: DtrItem[], platformSelection: number, mode: number) {
```
Comment (bhollis) [inline src/app/destinyTrackerApi/d2-bulkFetcher.ts]: It'd be nice to type `platformSelection` as `DestinyMembershipType` and `mode` as whatever it is

File: src/app/item-review/destiny-tracker.service.ts [hunk 4]
```diff
@@ -199,6 +207,10 @@ export function DestinyTrackerService(
       return _d2bulkFetcher.getCache();
     },
 
+    getReviewMode(defs: D2ManifestDefinitions, review: DtrUserReview): string {
+      return getReviewMode(defs, review);
+    },
```
Comment (bhollis) [inline src/app/item-review/destiny-tracker.service.ts]: You can write this as `getReviewMode,` since you're just calling the function with the same name and same arguments.

File: src/app/item-review/item-review.component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bhollis) [reply src/app/item-review/item-review.component.js]: No idea. That's super weird.

File: src/app/item-review/item-review.component.js [hunk 5]
```diff
@@ -92,6 +97,20 @@ function ItemReviewController(dimDestinyTrackerService, $scope, $rootScope) {
 
   vm.reviewLabels = [5, 4, 3, 2, 1];
 
+  vm.defs = {};
+
+  getDefinitions().then((defs) => {
```
Comment (bhollis) [inline src/app/item-review/item-review.component.js]: I think this component also gets used for Destiny 1. Can we wrap this in a check that the item is a D2 item to avoid loading the D2 manifest in D1 mode?

File: src/app/item-review/item-review.component.js [hunk 6]
```diff
@@ -92,6 +97,20 @@ function ItemReviewController(dimDestinyTrackerService, $scope, $rootScope) {
...
+    vm.reviewModeOptions = [
+      { key: 0, value: $i18next.t('DtrReview.ModeNotSpecified') },
+      { key: 7, value: defs.ActivityMode[1164760493].displayProperties.name },
+      { key: 5, value: defs.ActivityMode[1164760504].displayProperties.name },
+      { key: 4, value: defs.ActivityMode[2043403989].displayProperties.name },
+      { key: 39, value: defs.ActivityMode[1370326378].displayProperties.name }
+    ];
```
Comment (bhollis) [inline src/app/item-review/item-review.component.js]: Note: you could also turn `getReviewMode` into an Angular filter rather than repeating these here

File: src/app/rating-mode/rating-mode.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+export default class RatingMode extends React.Component<Props, State> {
+  private dropdownToggler: HTMLElement | null;
+
+  private reviewModeOptions = [
+    { key: 0, text: t('DtrReview.ModeNotSpecified') },
+    { key: 7, text: this.props.defs.ActivityMode[1164760493].displayProperties.name },
+    { key: 5, text: this.props.defs.ActivityMode[1164760504].displayProperties.name },
+    { key: 4, text: this.props.defs.ActivityMode[2043403989].displayProperties.name },
+    { key: 39, text: this.props.defs.ActivityMode[1370326378].displayProperties.name }
+  ];
```
Comment (bhollis) [inline src/app/rating-mode/rating-mode.tsx]: Or just extract them into a function or something?

File: src/app/rating-mode/rating-mode.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    if (!this.props.defs) {
+      return;
+    }
+
+    return (
+      <div>
+        <ClickOutside onClickOutside={this.closeDropdown}>
+          <span className="link" onClick={this.toggleDropdown} title={t('DtrReview.ForGameMode')}>
+            <i className={classNames('fa', 'fa fa-thumbs-up')}/>
```
Comment (bhollis) [inline src/app/rating-mode/rating-mode.tsx]: I don't think you need `classNames` for a static list of classes. Just use the string directly.

File: src/app/settings/settings.component.ts [hunk 9]
```diff
@@ -73,6 +74,16 @@ export function SettingsController(
...
 
+  getDefinitions().then((defs) => {
+    vm.reviewModeOptions = [
+        { key: 0, value: $i18next.t('DtrReview.ModeNotSpecified') },
+        { key: 7, value: defs.ActivityMode[1164760493].displayProperties.name },
+        { key: 5, value: defs.ActivityMode[1164760504].displayProperties.name },
+        { key: 4, value: defs.ActivityMode[2043403989].displayProperties.name },
+        { key: 39, value: defs.ActivityMode[1370326378].displayProperties.name }
+      ];
+  });
```
Comment (bhollis) [inline src/app/settings/settings.component.ts]: Starting to look familiar 😂

File: src/app/shell/Header.tsx [hunk 10]
```diff
@@ -90,10 +96,21 @@ export default class Header extends React.Component<Props, State> {
...
-      showSearch: false
+      showSearch: false,
+      defs: undefined
     };
   }
 
+  async getDefinitions() {
+    const defs = await getDefinitions();
+
+    if (defs.eagerTablesLoaded) {
```
Comment (bhollis) [inline src/app/shell/Header.tsx]: I'm not sure I understand the point of this block - what is it doing?

File: src/app/shell/Header.tsx [hunk 11]
```diff
@@ -214,6 +233,7 @@ export default class Header extends React.Component<Props, State> {
         </div>
 
         <span className="header-right">
+          {(!showSearch && account && account.destinyVersion === 2 && settings.showReviews && defs && defs.ActivityMode) && <RatingMode defs={defs} D2StoresService={this.props.D2StoresService} />}
```
Comment (bhollis) [inline src/app/shell/Header.tsx]: Why would `defs` be available but not have `ActivityMode`?

File: src/app/destiny2/d2-definitions.service.ts (hunk 2 above)
Comment (48klocs) [reply src/app/destiny2/d2-definitions.service.ts]: As part of this, you'll see that `header.tsx` does a polling loop on `getDefinitions`.

What I found was (and this doesn't make a whole lot of sense to me either), when you await definitions early enough, you get back a version of defs with the lazy loaded tables, but not the eager-loaded tables. Given that the underlying promise doesn't get returned until the eager tables are attached, this seems like something that shouldn't happen.

I put that in as a semaphore after the eager loading block to signal that the eager tables are really loaded in.

As far as the funky array assignment, `[ts] Property 'eagerTablesLoaded' does not exist on type '{}'.`

I was back in on the header (you mentioned item reviews could be pulling in D2 manifest for D1 users, and the same problem existed there) and waiting for the account to load seems to negate the race condition that existed here, so I dumped the new property.

Comment (bhollis) [reply src/app/destiny2/d2-definitions.service.ts]: Hmm, that behavior you're describing is counterintuitive enough that I'd rather figure out exactly why you're seeing that, than work around it.

File: src/app/destinyTrackerApi/reviewModesFetcher.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+export enum DtrActivityModes {
+  notSpecified = 0,
+  playerVersusEnemy = 7,
+  playerVersusPlayer = 5,
+  raid = 4,
+  trials = 39
+}
+
+export function getReviewModes(defs?: D2ManifestDefinitions): D2ReviewMode[] {
+  if (!defs || !defs.ActivityMode) {
```
Comment (bhollis) [inline src/app/destinyTrackerApi/reviewModesFetcher.ts]: I don't think there'd be a case where `defs.ActivityMode` isn't there. .

File: src/app/destinyTrackerApi/reviewModesFetcher.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  raid = 2043403989,
+  trials = 1370326378
+}
+
+export enum DtrActivityModes {
+  notSpecified = 0,
+  playerVersusEnemy = 7,
+  playerVersusPlayer = 5,
+  raid = 4,
+  trials = 39
```
Comment (bhollis) [inline src/app/destinyTrackerApi/reviewModesFetcher.ts]: You can express these in terms of their enums, like: `trials = DestinyActivityModeType.TrialsOfTheNine`. You may not even need this.

File: src/app/destinyTrackerApi/reviewModesFetcher.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import { t } from 'i18next';
+import { D2ManifestDefinitions } from "../destiny2/d2-definitions.service";
+
+export interface D2ReviewMode {
+  mode: number;
+  description: string;
+}
+
+export enum ActivityModeHashes {
```
Comment (bhollis) [inline src/app/destinyTrackerApi/reviewModesFetcher.ts]: Don't export this

File: src/app/destinyTrackerApi/reviewModesFetcher.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import { t } from 'i18next';
+import { D2ManifestDefinitions } from "../destiny2/d2-definitions.service";
+
+export interface D2ReviewMode {
```
Comment (bhollis) [inline src/app/destinyTrackerApi/reviewModesFetcher.ts]: No need to export this

File: src/app/destinyTrackerApi/reviewModesFetcher.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+export enum ActivityModeHashes {
+  playerVersusEnemy = 1164760493,
+  playerVersusPlayer = 1164760504,
+  raid = 2043403989,
+  trials = 1370326378
+}
+
+export enum DtrActivityModes {
```
Comment (bhollis) [inline src/app/destinyTrackerApi/reviewModesFetcher.ts]: Don't export this either

File: src/app/item-review/destiny-tracker.service.ts [hunk 17]
```diff
@@ -44,6 +44,8 @@ export interface DtrUserReview {
   isIgnored?: boolean;
   selectedPerks: number[];
   attachedMods: number[];
+  mode: number;  // play mode; corresponds to DestinyActivityModeType
+  sandbox: number; // sandbox season (1 was the first, 2 is the March 2018 "go fast" update)
```
Comment (bhollis) [inline src/app/item-review/destiny-tracker.service.ts]: You can type this as `1 | 2`

File: src/app/item-review/destiny-tracker.service.ts [hunk 18]
```diff
@@ -44,6 +44,8 @@ export interface DtrUserReview {
   isIgnored?: boolean;
   selectedPerks: number[];
   attachedMods: number[];
+  mode: number;  // play mode; corresponds to DestinyActivityModeType
```
Comment (bhollis) [inline src/app/item-review/destiny-tracker.service.ts]: Should this be `DestinyActivityModeType` then?

File: src/app/item-review/item-review.component.js [hunk 19]
```diff
@@ -209,6 +232,14 @@ function ItemReviewController(dimDestinyTrackerService, $scope, $rootScope) {
     }
   };
 
+  vm.translateReviewMode = function(review) {
+    if (!vm.defs || !vm.defs.ActivityMode) {
+      getDefinitions().then((defs) => { translateReviewMode(defs, review); });
+    }
+
+    return translateReviewMode(vm.defs, review);
+  };
```
Comment (bhollis) [inline src/app/item-review/item-review.component.js]: You should just always use the promise version. It'll return instantly if it's already resolved.

File: src/app/item-review/item-review.component.js [hunk 20]
```diff
@@ -209,6 +232,14 @@ function ItemReviewController(dimDestinyTrackerService, $scope, $rootScope) {
     }
   };
 
+  vm.translateReviewMode = function(review) {
+    if (!vm.defs || !vm.defs.ActivityMode) {
+      getDefinitions().then((defs) => { translateReviewMode(defs, review); });
```
Comment (bhollis) [inline src/app/item-review/item-review.component.js]: The braces aren't necessary.

File: src/app/rating-mode/rating-mode.tsx [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  private modeChange = (e?) => {
+    if (!e || !e.target) {
+      return;
+    }
+
+    const newModeSelection = e.target.value;
+    settings.reviewsModeSelection = newModeSelection;
+    settings.reviewsModeSelection = newModeSelection;
```
Comment (bhollis) [inline src/app/rating-mode/rating-mode.tsx]: These two lines look the same.

File: src/app/item-review/destiny-tracker.service.ts (hunk 18 above)
Comment (48klocs) [reply src/app/item-review/destiny-tracker.service.ts]: No; this is data that the DTR API returns.

Which is a number that corresponds to a `modeType` in `DestinyActivityModeType`, but if you look at the values in the manifest, that's not going to get you a unique entry.

File: src/app/item-review/item-review.component.js (hunk 19 above)
Comment (48klocs) [reply src/app/item-review/item-review.component.js]: `"[$rootScope:infdig] 10 $digest() iterations reached. Aborting!
Watchers fired in the last 5 iterations: []
http://errors.angularjs.org/1.6.9/$rootScope/infdig?p0=10&p1=%5B%5D"`

I must be doing something wrong in there?

Comment (bhollis) [reply src/app/item-review/item-review.component.js]: Oh, I see what's happening. You should never call a function from ng-bind or other watched expressions. Use a filter instead, or preprocess these values.

Comment (48klocs) [reply src/app/item-review/item-review.component.js]: Ah, that makes sense to me.

It’s using the same array of values processed out of the manifest, but that only needs to happen once. 

Is there an expected place to hold (lazy loaded) global state? I don’t know if we have any singletons in the app.

File: src/app/item-review/destiny-tracker.service.ts (hunk 17 above)
Comment (48klocs) [reply src/app/item-review/destiny-tracker.service.ts]: Doesn't that constrain it to being 1 or 2?

I'd rather not have to push a new app version because the API incremented the sandbox.

File: src/app/destinyTrackerApi/reviewModesFetcher.ts (hunk 13 above)
Comment (48klocs) [reply src/app/destinyTrackerApi/reviewModesFetcher.ts]: I went with the enumerated type so that we're not dealing with magic numbers internally for the agreed-upon mode numbers that DTR's OK with receiving (and we're expecting).

Internally, we can point to `DestinyActivityModeType` for most, but this falls down on `notSpecified` (there's a few modes where the type is 0 in the manifest), and with the data that we're receiving from the API, we don't get a mode (but we could certainly perform a lookup and post-process an object in).

If it would make more sense, I could move this over to the tracker service, so I could have `DtrUserReview`'s `mode` property of this type (and tighten the interfaces on methods in there from accepting `mode` as a `number` to accepting it as a `DtrActivityMode`.
